w11h03/test/pgdp/networking/DataHandlerTest.java

```diff
@@ -0,0 +1,128 @@
+        assertEquals("POST", actual.requestMethod());
+        assertEquals(List.of("application/json"), actual.requestHeaders().get("Content-Type"));
+        assertEquals(URI.create("http://carol.sse.cit.tum.de/api/user/register"), actual.requestURI());
+        assertEquals("{\"username\": \"" + username + "\",\"tum_kennung\": \"" + id + "\"}", actual.getBodyUTF8());
+    }
+
+    @Test
+    public void registerSuccessTest() {
+        Pair<Boolean, Request> result = testEndpoint("/api/user/register", 200, "", () -> dataHandler.register(USERNAME, ID)); // TODO: https://zulip.in.tum.de/#narrow/stream/1525-PGdP-W11H03/topic/Definition.20erfolgreiche.20Anfragen
+        registerRequestTest(result.getRight(), USERNAME, ID);
+        assertTrue(result.getLeft());
+    }
+
+    @Test
+    public void registerFailTest() {
+        Pair<Boolean, Request> result = testEndpoint("/api/user/register", 422, "", () -> dataHandler.register(USERNAME, ID)); // TODO: https://zulip.in.tum.de/#narrow/stream/1525-PGdP-W11H03/topic/Definition.20erfolgreiche.20Anfragen
+        registerRequestTest(result.getRight(), USERNAME, ID);
+        assertFalse(result.getLeft());
+    }
+
+    private void requestTokenRequestTest(Request actual, String username, String password) {
+        assertEquals("POST", actual.requestMethod());
+        assertEquals(List.of("application/x-www-form-urlencoded"), actual.requestHeaders().get("Content-Type"));
+        assertEquals(URI.create("http://carol.sse.cit.tum.de/token"), actual.requestURI());
+        assertEquals("username=" + username + "&password=" + password, actual.getBodyUTF8());
+    }
+
+    @Test
+    public void requestTokenSuccessTest() {
+        Pair<String, Request> result = testEndpoint("/token", 200, "{\n" +
+                "  \"access_token\": \"" + DUMMY_TOKEN + "\",\n" +
+                "  \"token_type\": \"someTokenType\"\n" +
+                "}", () -> callRequestToken(dataHandler, USERNAME, PASSWORD));
+
+        requestTokenRequestTest(result.getRight(), USERNAME, PASSWORD);
+
+        assertEquals(DUMMY_TOKEN, result.getLeft());
+    }
+
+    private <T> Pair<T, Request> testEndpoint(String endpoint, int code, String responseBody, Supplier<T> fn) {
+        AtomicReference<Request> req = new AtomicReference<>();
+
+        server.createContext(endpoint, exchange -> {
+            req.set(new Request(exchange.getRequestMethod(), exchange.getRequestHeaders(), exchange.getRequestURI(), exchange.getRequestBody().readAllBytes()));
+
+            byte[] body = responseBody.getBytes(StandardCharsets.UTF_8);
+            exchange.sendResponseHeaders(code, body.length);
+            exchange.getResponseBody().write(body);
+            exchange.close();
+        });
+
+        T result = fn.get();
+
+        server.removeContext(endpoint);
+
+        return Pair.of(result, Objects.requireNonNull(req.get(), "provided function did not send request to specified endpoint"));
+    }
```

Comment:
could you make this method public so other files can use it too?

Reply:
Latest commit should more or less resolve this